### ISSUE TITLE
eslint.linter API exposes global mutable state

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#8454]
## Problem

ESLint exports a `linter` object, which is intended to allow libraries and integrations to lint arbitrary pieces of text. However, the `linter` object is mutable, and clients need to mutate global state in order to use it effectively. For example, in order to define a custom rule, clients need to use `eslint.linter.defineRule()`, which adds the rule for *every* module using eslint in the current process. This can lead to confusing errors if multiple clients are using the ESLint API simultaneously, because they inadvertently interfere with each other.

Examples of cases where this can pose a problem:

* Two editor integrations running ESLint simultaneously can interfere with each other
* An ESLint rule that runs a separate ESLint instance (e.g. when "linting" an ESLint config file to verify that it doesn't cause new errors in other projects) will inadvertently break the ESLint instance that it's running.

Previous discussions: https://github.com/eslint/eslint/issues/205, https://github.com/eslint/eslint/issues/395#issuecomment-28862653

## Proposal

We should export an `eslint.Linter` API. `Linter` would be a constructor for objects like `eslint.linter`, except that different instances of `Linter` would be independent of each other.

```js
// previous API usage:

const eslint = require("eslint");

// this will define a rule for *everyone* using the eslint module
eslint.linter.defineRule("foo", rule);

eslint.linter.verify(text, config);
```

```js
// new API usage

const eslint = require("eslint");
const linter = new eslint.Linter();

// this will only define a rule for this particular linter instance
linter.defineRule("foo", rule);
```

We could make this change backwards compatible by creating an instance of `Linter` internally and continuing to export it as `eslint.linter`. However, we would deprecate `eslint.linter` in favor of `eslint.Linter`. Additionally, `linter.reset` would be deprecated in favor of simply creating a new `Linter` instance.

I think this proposal would make it much easier to use ESLint's Node.js API without creating unexpected errors for other clients that use the API simultaneously.

**Comment by gyandeeps** [comment]
Based on the votes above, I am marking it as accepted. I think we should do this with 4.0, thoughts?

**Comment by not-an-aardvark** [comment]
I think this should probably go through the TSC because it's an enhancement to core.

> I think we should do this with 4.0, thoughts?

Assuming this is a semver-minor change, I don't think we should add it to 4.0 because we already have enough to do for 4.0, and we could always add it in another release afterwards.

**Comment by mysticatea** [comment]
Definitely :+1:.

**Comment by kaicataldo** [comment]
As discussed in today's TSC meeting, this has now been accepted as part of the v4.0 release.
